Scripts/Winwing/prosim_a320_winwing_cdu.py

```diff
@@ -0,0 +1,384 @@
+    return json.dumps(message, separators=(',', ':')) 
+
+class ProSimClient:
+    pyprosim: PyProsim = None
+
+    def __init__(self) -> None:
+        try:
+            self.pyprosim = PyProsim(Path("ProSimSDK.dll").absolute())
+            self.connected = False
+        except Exception as e:
+            logging.error(f"Failed to initialize ProSim SDK: {e}")
+            raise
+
+    def connect(self) -> bool:
+        try:
+            logging.info(f"Connecting to ProSim at {host}")
+            self.pyprosim.connect(host, synchronous=True)
+            self.connected = True
+            logging.info("Successfully connected to ProSim")
+            return True
+        except Exception as e:
+            logging.error(f"Failed to connect to ProSim: {e}")
+            self.connected = False
+            return False
+
+    def add_callback(self, dataref_name: str, callback: Callable) -> bool:
+        try:
+            self.pyprosim.add_dataref(dataref_name, 50, callback)
+            return True
+        except Exception as e:
+            logging.error(f"Failed to add callback for {dataref_name}: {e}")
+            return False
+
+    def get_dataref_value(self, dataref_name: str):
+        try:
+            return self.pyprosim.get_dataref_value(dataref_name)
+        except Exception as e:
+            logging.error(f"Error getting dataref value for {dataref_name}: {e}")
+            return None
+
+
+class ProSimCDUClient:
+    def __init__(self, prosim_client: ProSimClient, websocket_uri: str, cdu_name: str, cdu_dataref_name: str) -> None:
+        self.prosim_client: ProSimClient = prosim_client
+        self.mobiflight: MobiFlightClient = MobiFlightClient(websocket_uri)
+        self.event_loop: Optional[asyncio.AbstractEventLoop] = None
+        self.cdu_name: str = cdu_name
+        self.cdu_dataref_name: str = cdu_dataref_name
+
+    def failed_to_connect(self) -> bool:
+        return self.mobiflight.retries >= self.mobiflight.max_retries
+
+    def setup_prosim(self) -> bool:
+        try:
+            self.prosim_client.connect()
+                    
+            # Check that we can access datarefs
+            datarefs = self.prosim_client.pyprosim.get_all_avail_datarefs()
+            if not datarefs:
+                logging.error(f"No datarefs found in ProSim for {self.cdu_name}")
+                return False
+                
+            self.prosim_client.add_callback(self.cdu_dataref_name, None)
+
+            logging.info(f"ProSim connection initialized for {self.cdu_name}")
+            return True
+        except Exception as e:
+            logging.error(f"ProSim setup failed for {self.cdu_name}: {e}")
+            return False
+
+    async def process_prosim(self) -> None:
+        while True:
+            dataref = self.prosim_client.get_dataref_value(self.cdu_dataref_name)
+            if dataref:
+                asyncio.run_coroutine_threadsafe(self.mobiflight.send(create_mobi_json(dataref)), self.event_loop)
```

Comment:
Since we are polling, I would propose to store the current value and only send to mobiflight if the value has changed. There is also a check before sending to the winwing device, but not sending at all to mobiflight would be the best. I assume each 100ms the same value is returned.

Reply:
Yep exactly, if you check the 737 implementation, I have exactly that, where I store the last value sent and only send if there is a diff

Reply:
The issue with the callback solution is that pythonnet runs it on the C# side, and it has problems interoping with asyncio. I was never able to get the callback solution to work nicely, so I think polling with checking should be good enough most likely